sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/MetricStorageRegistry.java

```diff
@@ -5,71 +5,88 @@
 
 package io.opentelemetry.sdk.metrics.internal.state;
 
+import io.opentelemetry.api.internal.GuardedBy;
 import io.opentelemetry.sdk.metrics.internal.descriptor.MetricDescriptor;
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
+import java.util.List;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentMap;
+import java.util.logging.Level;
+import java.util.logging.Logger;
 
 /**
- * Responsible for storing metrics (by name) and returning access to input pipeline for instrument
- * wiring.
+ * Responsible for storing metrics by {@link MetricDescriptor} and returning access to input
+ * pipeline for instrument measurements.
  *
- * <p>The rules of the registry:
- *
- * <ul>
- *   <li>Only one storage type may be registered per-name. Repeated look-ups per-name will return
- *       the same storage.
- *   <li>The metric descriptor should be "compatible", when returning an existing metric storage,
- *       i.e. same type of metric, same name, description etc.
- *   <li>The registered storage type MUST be either always Asynchronous or always Synchronous. No
- *       mixing and matching.
- * </ul>
+ * <p>Each descriptor in the registry results in an exported metric stream. Under normal
+ * circumstances each descriptor shares a unique {@link MetricDescriptor#getName()}. When multiple
+ * descriptors share the same name, an identity conflict has occurred. The registry detects identity
+ * conflicts on {@link #register(MetricStorage)} and logs diagnostic information when they occur.
+ * See {@link MetricDescriptor#isCompatibleWith(MetricDescriptor)} for definition of compatibility.
  *
  * <p>This class is internal and is hence not for public use. Its APIs are unstable and can change
  * at any time.
  */
 public class MetricStorageRegistry {
-  // TODO: Maybe we store metrics *and* instrument interfaces separately here...
-  private final ConcurrentMap<String, MetricStorage> registry = new ConcurrentHashMap<>();
+  private static final Logger logger = Logger.getLogger(MetricStorageRegistry.class.getName());
+
+  private final Object lock = new Object();
+
+  @GuardedBy("lock")
+  private final ConcurrentMap<MetricDescriptor, MetricStorage> registry = new ConcurrentHashMap<>();
 
   /**
    * Returns a {@code Collection} view of the registered {@link MetricStorage}.
    *
    * @return a {@code Collection} view of the registered {@link MetricStorage}.
    */
   public Collection<MetricStorage> getMetrics() {
-    return Collections.unmodifiableCollection(new ArrayList<>(registry.values()));
+    synchronized (lock) {
+      return Collections.unmodifiableCollection(new ArrayList<>(registry.values()));
```

Comment:
Old code so no worries, but noticed that the javadoc says "view" even though we copy